Controllers/users.js

```diff
@@ -1,10 +1,34 @@
+const bcrypt = require("bcryptjs");
+const { isEmail, isStrongPassword } = require("validator");
+
 const User = require("../Models/users");
 
 const signin = async (req, res) => {
   res.json({ message: "Signed in!" });
 };
+
 const signup = async (req, res) => {
-  res.json({ message: "Signed up!" });
+  const { email, password } = req.body;
+
+  try {
+    if (!email) throw Error("Email is required!");
+    if (!isEmail(email)) throw new Error("Email is not valid.");
+
+    if (!password) throw Error("Password is required!");
+    if (!isStrongPassword(password)) throw new Error("Password is not strong.");
+
+    let user = await User.findOne({ email });
+
+    if (user) throw Error("Email address is already used!");
+
+    const hash = await bcrypt.hash(password, 10);
```

Comment:
I don't know how the `hashing` is working with this library, but you might be considering providing a self created `salt` key, so the hashing process using this salt and produces always the same output, when giving the same information. Right now, it seems that the hashed password is always different, because you say, pls use a salt I don't care about. That said, I don't know how you plan the comparing process, but how I said, could be a problem later on without a fixed salt.

Reply:
Yes, that made me think, too. In the signin function I will use the `bycript.compare(hashedPasswordFromDB, signinPassword)`

The interesting thing here is, bycript uses the salt in the hashing process and then concatenates the salt to the hash result.
Here are some console.logs I deleted before pushing:

```javascript
const passwordToHash = "someRandomPassword"
const salt = await bcrypt.genSalt(10);
console.log("salt: ", salt);
const hashedPassword = await bcrypt.hash(passwordToHash, salt)
console.log("hashed Password: ", hashedPassword);
```


salt: 
**$2a$10$6sEOkplEcFUiPIp4nHdd5O**
hashed Password: 
**$2a$10$6sEOkplEcFUiPIp4nHdd5O**Ta1CkujElJpFwqwB0DzhVfO50RrHyge

bcrypt knows which part is the salt, uses it to hash the signin password and then compares the result. 👍